asllib/StaticModel.ml

```diff
@@ -2,366 +2,428 @@ open AST
+  (** A map from a monomial. *)
+
+  type t = Q.t MonomialMap.t
+  (** A polynomial.
+      Maps each monomial to its factor.
+      For example, {m X^2 - X + 4 } is represented by
+      {m X^2 \to 1, X \to -1, 1 \to 4 } *)
 
-  (* Available from 4.11.0 *)
-  let filter_map f m =
-    fold
-      (fun key v r -> match f key v with None -> r | Some v -> add key v r)
-      m empty
+  let compare = MonomialMap.compare Q.compare
+
+  let single_term mono factor =
+    if Q.equal factor Q.zero then MonomialMap.empty
+    else MonomialMap.singleton mono factor
+
+  let to_mono poly =
+    if MonomialMap.cardinal poly = 1 then Some (MonomialMap.choose poly)
+    else None
+
+  let add poly1 poly2 =
+    MonomialMap.union
+      (fun _ c1 c2 ->
+        let coeff = Q.add c1 c2 in
+        if Q.equal coeff Q.zero then None else Some coeff)
+      poly1 poly2
+
+  let scale factor poly =
+    assert (factor <> Q.zero);
+    MonomialMap.map (Q.mul factor) poly
+
+  let termwise f poly =
+    MonomialMap.fold
+      (fun mono factor -> add (f factor mono))
+      poly MonomialMap.empty
+
+  let mult_mono poly factor mono =
+    termwise
+      (fun f m -> single_term (Monomial.mult m mono) (Q.mul f factor))
+      poly
+
+  let neg poly = MonomialMap.map Q.neg poly
+  let mult poly1 poly2 = termwise (mult_mono poly1) poly2
+
+  let divide_by_term poly factor mono =
+    termwise
+      (fun f m -> single_term (Monomial.divide m mono) (Q.div f factor))
+      poly
+
+  let extract_constant_term poly =
+    let c = try MonomialMap.find Monomial.one poly with Not_found -> Q.zero
+    and p = MonomialMap.remove Monomial.one poly in
+    (c, p)
+
+  let is_constant poly =
+    if MonomialMap.is_empty poly then Some Q.zero
+    else if MonomialMap.cardinal poly = 1 then
+      MonomialMap.find_opt Monomial.one poly
+    else None
+
+  let to_expr poly =
+    List.fold_left
+      (fun acc (m, c) ->
+        if acc = Expr.zero then Monomial.to_expr m c
+        else
+          let e_m = Monomial.to_expr m (Q.abs c) in
+          Expr.add acc (Q.sign c, e_m))
+      Expr.zero
+      (MonomialMap.bindings poly |> List.rev)
+
+  let pp f poly =
+    let open Format in
+    if MonomialMap.is_empty poly then pp_print_string f "0"
+    else (
+      pp_open_hvbox f 2;
+      let pp_sep f () = fprintf f "@ + " in
+      PP.pp_print_seq ~pp_sep Monomial.pp_with_factor f
+        (MonomialMap.to_seq poly);
+      pp_close_box f ())
 end
 
-(** A polynomial.
+module Conjunction : sig
+  type eq = Zero | NonZero
+  type t
 
-      For example, {m X^2 - X + 4 } is represented by
-      {m X^2 \to 1, X \to -1, 1 \to 4 } *)
-type polynomial = Sum of Q.t MMap.t  (** Maps each monomial to its factor. *)
+  val empty : t
+  val is_bottom : t -> bool
+  val is_empty : t -> bool
+  val of_bool : bool -> t
+  val single_conjunct : Polynomial.t -> eq -> t
+  val conj : t -> t -> t
+  val to_expr : t -> expr option
+  val is_trivial : t -> bool option
```

Comment:
I don't understand what this does. Why does it return a `bool option` and not just `bool`?

Reply:
I've added an `ocamldoc` comment to explain: if the conjunction is trivially true or false, this returns `Some true` or `Some false`. If not, then it returns `None`.

Reply:
It feels like instead of a comment, this needs to be a type. perhaps something like:
`TriviallyTrue | TriviallyFalse | Nontrivial`.

Reply:
I've introduced this type as you suggest!